impress-org/givewp
# Feature: Introduce Undo/Redo for Form State

## Description



This PR introduces Undo and Redo functionality using `useUndoableReducer` (`use-undoable-reducer`).

This is similar to the implementation of History Management in the Isolated Block Editor, which uses `redux-undo`. However, we are using the Context API, which works better with a simple undoable reducer. That said, both packages follow the [Implementing Undo History recipe](https://redux.js.org/usage/implementing-undo-history) documented by Redux.

Note: Most of the change set involves combining the `editor-state` store with the `form-state` store. This was done to include `EditorMode` state in the "undoable" reducer, which improves the experience when changes are made in multiple "modes" (ie Design, the Edit Form).

## Affects



- Adds `canUndo`, `canRedo` to form state (similar to `isDirty`).
- Merges the `editor-state` store with the `form-state` store.
- Removes the `<EditorStateProvider />` HOC.
- Moves unique field name logic to middleware.

Note: Field name uniqueness was moved to a middleware to avoid a side effect which triggered additional state changes for a single action. For example, duplicating a field would result in two state changes: add field, update metaUUID. However, undoing the "update metaUUID" state change would unset the metaUUID which re-trigger the side effect to update the metaUUID, resulting in an infinite loop. With this PR the middleware approach squashes the two changes resulting in an atomic changeset.

## Visuals



### Undo form field movement
![Peek 2023-10-10 11-15](https://github.com/impress-org/givewp/assets/10858303/8f5a10cd-3083-41b7-88db-295857653601)

### Undo form design changes
![Peek 2023-10-10 11-16](https://github.com/impress-org/givewp/assets/10858303/be0b1444-66d4-41b9-902f-85646ec50c58)

### Undo/Redo across Editor Modes
![Peek 2023-10-10 11-20](https://github.com/impress-org/givewp/assets/10858303/a6fe7eab-7d48-4aae-8c7c-570c94040673)

## Testing Instructions

- Changing the label for new fields changes the key
- Duplicating a field also checks the meta key
- New fields' meta keys are editable
- Existing fields' meta keys are locked (an can be unlocked)
- Certain keywords are prevented

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205502083833538
